official/resnet/resnet_model.py

```diff
@@ -205,162 +205,99 @@ def projection_shortcut(inputs):
 
   return tf.identity(inputs, name)
 
 
-def cifar10_resnet_v2_generator(resnet_size, num_classes, data_format=None):
-  """Generator for CIFAR-10 ResNet v2 models.
+class Model(object):
+  """Base class for building the Resnet v2 Model.
+  """
+  resnet_size = None
+  num_classes = None
+  data_format = None
+  input_shape = None
 
-  Args:
-    resnet_size: A single integer for the size of the ResNet model.
-    num_classes: The number of possible classes for image classification.
-    data_format: The input format ('channels_last', 'channels_first', or None).
-      If set to None, the format is dependent on whether a GPU is available.
+  params = None
 
-  Returns:
-    The model function that takes in `inputs` and `is_training` and
-    returns the output tensor of the ResNet model.
-
-  Raises:
-    ValueError: If `resnet_size` is invalid.
-  """
-  if resnet_size % 6 != 2:
-    raise ValueError('resnet_size must be 6n + 2:', resnet_size)
+  def __init__(self, resnet_size, data_format=None):
+    """Creates a model for classifying an image.
 
-  num_blocks = (resnet_size - 2) // 6
+    Args:
+      resnet_size: A single integer for the size of the ResNet model.
+      data_format: Input format ('channels_last', 'channels_first', or None).
+        If set to None, the format is dependent on whether a GPU is available.
+    """
+    self.resnet_size = resnet_size
 
-  if data_format is None:
-    data_format = (
+    if not data_format:
+      data_format = (
         'channels_first' if tf.test.is_built_with_cuda() else 'channels_last')
 
-  def model(inputs, is_training):
-    """Constructs the ResNet model given the inputs."""
     if data_format == 'channels_first':
-      # Convert the inputs from channels_last (NHWC) to channels_first (NCHW).
-      # This provides a large performance boost on GPU. See
-      # https://www.tensorflow.org/performance/performance_guide#data_formats
-      inputs = tf.transpose(inputs, [0, 3, 1, 2])
+      self.input_shape = [0, 3, 1, 2]
+    else:
+      assert data_format == 'channels_last'
+      self.input_shape = [0, 1, 2, 3]
 
-    inputs = conv2d_fixed_padding(
-        inputs=inputs, filters=16, kernel_size=3, strides=1,
-        data_format=data_format)
-    inputs = tf.identity(inputs, 'initial_conv')
+    self.data_format = data_format
 
-    inputs = block_layer(
-        inputs=inputs, filters=16, block_fn=building_block, blocks=num_blocks,
-        strides=1, is_training=is_training, name='block_layer1',
-        data_format=data_format)
-    inputs = block_layer(
-        inputs=inputs, filters=32, block_fn=building_block, blocks=num_blocks,
-        strides=2, is_training=is_training, name='block_layer2',
-        data_format=data_format)
-    inputs = block_layer(
-        inputs=inputs, filters=64, block_fn=building_block, blocks=num_blocks,
-        strides=2, is_training=is_training, name='block_layer3',
-        data_format=data_format)
+    self.params = self.get_model_params()
```

Comment:
Would it make sense for all params to be arguments to `__init__` instead of this separate `get_model_params()` method?

Reply:
`get_model_params` makes it easy to subclass this for each dataset and just define the params for model setup in the class. Passing through init would work, but makes it harder to run some of the checks that are easy to include in a class. Six of one, half dozen of another at the end of the day, but since we're going class-based, it seems more Pythonic to keep it that way for the specific versions as well.

Reply:
Didn't quite follow. I was imagining something like this:

```python
class Model(object):
   def __init__(self,
                       resnet_size,
                       num_classes,
                       num_filters,
                       kernel_size,
                       ...):
     # ...

class Cifar10Model(Model):
   def __init__(self, resnet_size, data_format=None):
         super(ImageNetModel, self).__init__(
                 resnet_size=resnet_size,
                 data_format=data_format,
                 num_classes=10,
                 kernel_size=3,
                 ...):
```

This way there is one clear specification of all the parameters in the `Model` class. With the `get_model_parameters()` method, which returns a dictionary, there is no obvious spec of what the keys in the dictionary should be. One has to comb through the implementation of the base class to figure that out.

I feel having that clear spec will be helpful.

Reply:
Oh, got it, I misunderstood. Yes, that makes sense-- will do.